### PR TITLE
texlive.combine: expose licensing information of combined packages [v2]

### PR DESCRIPTION
**Pull request by apfelkuchen6**
~~This is a slight modification of fa7428470abdfea7a4e1e1ac5d3e9ebac452f2a2, which has been reverted in 205ee073b053fc4d87d5adf2ebd44ebbef7bca4d due to it breaking the evaluation on Hydra (probably because of excessive stack usage). Sorry about that.~~

~~In this version, `lib.foldl` has been replaced by `builtins.foldl'` in both folds, which does indeed cut down the neccessary stack size significantly: `nix-instantiate -A texlive.combined.scheme-full` runs successfully with the stack limited to 300kB (`ulimit -s 300`), whereas the version with `lib.foldl` required more than 20MB.~~

~~I could not measure a significant difference in runtime between the versions without licensing metadata, `lib.foldl` and `builtins.foldl'`. All three versions need around 2.4s to instantiate `texlive.combined.scheme-full` on my machine with run-to-run deviations being larger than the differences between the versions.~~

~~Not sure how to correctly test this, as `nixpkgs-review` as well as @ofborg also evaluated nixpkgs in the previous version without problems.~~

This now uses hardcoded version information to reduce the evaluation hit further. A tests ensures that the hardcoded information is kept up to date.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by apfelkuchen6** [comment]
I was able to reproduce the crash. In `nix-shell -p hydra-unstable`, `hydra-eval-jobs '<nixpkgs/pkgs/top-level/release.nix>' -I nixpkgs=https://github.com/apfelkuchen6/nixpkgs/archive/refs/heads/texlive-licenses.zip` dies with `child process (2378940) killed by signal=11`. This no longer happens with this PR.

So I think this should be fine now.

**Comment by vcunat** [comment]
I haven't looked properly yet, but besides crashes, I'd probably also consider if it might still be very expensive (RAM + CPU).

**Comment by xworld21** [comment]
I tested it with some care. On my machine, concatenating the licenses adds some 10% time to the evaluation of `scheme-full`, which is already quite slow (3.5s).

I think the benefits do not justify the cost after all – once each package has its license, Nix will block the evaluation if there are unfree packages (which will become possible if/when we add tlcontrib to the repository). So lifting the licenses up to the combined outputs seems unnecessary.

**Comment by apfelkuchen6** [comment]
> I tested it with some care. On my machine, concatenating the licenses adds some 10% time to the evaluation of `scheme-full`, which is already quite slow (3.5s).

That's really interesting. On my machine, the increase is only about 1.5% on average when evaluating 100 times, with visually larger run-to-run variances:
```
❯ for i in {1..100}; do NIX_SHOW_STATS=1 nix-instantiate -A texlive.combined.scheme-full |& tail -n +3 | jq .cpuTime >> stats-with-meta; done
❯ git checkout @~
HEAD is now at 030ba16f5c5 Merge pull request #238698 from r-ryantm/auto-update/oh-my-zsh
❯ for i in {1..100}; do NIX_SHOW_STATS=1 nix-instantiate -A texlive.combined.scheme-full |& tail -n +3 | jq .cpuTime >> stats-without-meta; done
❯ paste -sd+ stats-without-meta | bc
144.6777937412261968
❯ paste -sd+ stats-with-meta | bc
146.8613368272781378
❯ bc -l <<< "146.8613368272781378 / 144.6777937412261968"
1.01509245496207574861
```
(https://gist.github.com/apfelkuchen6/7f0bb8ba3f9bb195c0a8661e21826880)

I agree that 10% slowdown is too much for a (mostly) cosmetic change, even if it only occurs with some setups.
 
> I think the benefits do not justify the cost after all – once each package has its license, Nix will block the evaluation if there are unfree packages (which will become possible if/when we add tlcontrib to the repository). So lifting the licenses up to the combined outputs seems unnecessary.

I still think that we should at least add some licensensing information to the combined schemes `texlive.combined.scheme-{full,medium,infraonly,context,basic}`. These actually end up in the channel metadata and on search.nixos.org. I don't think not having any licensing information there is acceptable, but a generic `licenses.free` should suffice. Now that we have the licensing information of the constituents available, we could also pre-generate the combined license lists for the schemes and hardcode them. What do you think?

There are some packages in nixpkgs with a lot of licensses (for example [labplot](https://search.nixos.org/packages?channel=23.05&from=0&size=50&sort=relevance&type=packages&query=labplot)), but nothing with as many licenses as texlive.combined.scheme-full. The complete license list should look somewhat like this on search.nixos.org:

licenses

**Comment by xworld21** [comment]
> I still think that we should at least add some licensensing information to the combined schemes `texlive.combined.scheme-{full,medium,infraonly,context,basic}`. These actually end up in the channel metadata and on search.nixos.org.

I suppose the proper solution is to expose the texlive packages for the world to see, which *might* happen before 23.11. Would that solve your issue?

In the meanwhile, I would support checking for 'unfree' and tagging the scheme free/unfree accordingly. If that is still slow, we can override the license on the default combined schemes, which are guaranteed to be free anyway; but we would be ready for tlcontrib and support local installations with non free packages (if they exist!).

**Comment by xworld21** [comment]
> Now that we have the licensing information of the constituents available, we could also pre-generate the combined license lists for the schemes and hardcode them. What do you think?

Seems reasonable if accompanied by a test, to catch changes.

**Comment by drupol** [comment]
Is this PR ready ?

**Comment by xworld21** [comment]
> Is this PR ready ?

Not yet, the current PR causes a hit on performance (maybe more pronounced on Darwin?) but @apfelkuchen6 made a good suggestion – precomputing the licenses – which gives the desired result without the evaluation hit.

**Comment by drupol** [comment]
Thanks for setting it in `draft`, it's much easier for reviewer :)

Feel free to ping me when it's ready for a quicker review.

File: pkgs/test/texlive/default.nix [hunk 1]
```diff
@@ -321,4 +321,46 @@
...
+        concatLicenses = builtins.foldl' (acc: el: if builtins.elem el acc then acc else acc ++ [ el ]);
+        savedLicenses = scheme: scheme.meta.license;
+        correctLicenses = scheme: builtins.foldl'
+                (acc: pkg: concatLicenses acc (lib.toList (pkg.meta.license or [])))
+                []
+                scheme.passthru.packages;
+        leq = (a: b: a.fullName <= b.fullName);
+        hasLicenseMismatch = scheme:
+          (lib.isDerivation scheme) &&
+          (lib.sort leq (savedLicenses scheme) != lib.sort leq (correctLicenses scheme));
```
Comment (xworld21) [inline pkgs/test/texlive/default.nix:342]:
```suggestion
          (lib.sort leq (savedLicenses scheme) != correctLicenses scheme);
```
Let's assert that the list is ordered.

File: pkgs/tools/typesetting/tex/texlive/default.nix [hunk 2]
```diff
@@ -264,13 +265,38 @@ in
 
     # Pre-defined combined packages for TeX Live schemes,
     # to make nix-env usage more comfortable and build selected on Hydra.
-    combined = with lib; recurseIntoAttrs (
+    combined = with lib;
+      let
+        # these license lists should be the union of the licenses of the packages the schemes contain.
```
Comment (xworld21) [inline pkgs/tools/typesetting/tex/texlive/default.nix:270]:
```suggestion
        # these license lists should be the sorted union of the licenses of the packages the schemes contain.
```

**Review by xworld21 — CHANGES_REQUESTED** [review]
LGTM but see suggestions. Could you please add a line to UPGRADING requesting to run the tests? We should document that the way to update licenses it to run the test, and on failure copy paste the error message.

**Comment by xworld21** [comment]
@GrahamcOfBorg build tests.texlive.licenses

**Review by xworld21 — APPROVED** [review]

**Review by drupol — APPROVED** [review]

**Comment by apfelkuchen6** [comment]
Rebased to solve conflicts.

@ofborg build tests.texlive

**Comment by drupol** [comment]
Can I merge this ?

**Comment by apfelkuchen6** [comment]
> Can I merge this ?

I'm not aware of any remaining issues, so I think this should be fine for merge.

**Pull request merged**